core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java

```diff
@@ -0,0 +1,283 @@
+import io.questdb.cairo.vm.api.MemoryCARW;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.BinarySequence;
+import io.questdb.std.FilesFacade;
+import io.questdb.std.FilesFacadeImpl;
+import io.questdb.std.Long256;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Mutable;
+import io.questdb.std.Numbers;
+import io.questdb.std.Vect;
+import io.questdb.std.str.Path;
+
+public class MemoryCARWSpillable implements MemoryCARW, Mutable {
+    private static final Log LOG = LogFactory.getLog(MemoryCARWSpillable.class);
+    // for non-mmap
+    private final int maxPages;
+    // flag
+    private long diskSpillThreshold;
+    private long extendSegmentMsb;
+    private FilesFacade ff = new FilesFacadeImpl();
+    private boolean isSpilled;
+    private MemoryCARW mem;
+    private long minMappedMemorySize = -1;
+    private int nativeMemoryTag;
+    private int opts;
+    // general
+    private long pageSize;
+    private int spillMemoryTag;
+    // for mmap
+    private Path spillPath;
+    private String tmpFile;
+
+    // TODO: For now assume the same pageSize
+    public MemoryCARWSpillable(long pageSize, int maxPages, long diskSpillThreshold, CairoConfiguration cairoConf){
+        this(pageSize, maxPages, diskSpillThreshold, MemoryTag.NATIVE_DEFAULT, MemoryTag.MMAP_DEFAULT);
+        this.spillPath = Path.getThreadLocal(cairoConf.getRoot()).concat(tmpFile).$();
+        ff.touch(spillPath);
+    }
+
+    public MemoryCARWSpillable(long pageSize, int maxPages, long diskSpillThreshold, Path path){
+        this(pageSize, maxPages, diskSpillThreshold, MemoryTag.NATIVE_DEFAULT, MemoryTag.MMAP_DEFAULT);
+        this.spillPath = path.concat(tmpFile).$();
+        ff.touch(path);
+    }
+
+    public MemoryCARWSpillable(long pageSize, int maxPages, long diskSpillThreshold, Path path, int nativeMemoryTag){
+        this(pageSize, maxPages, diskSpillThreshold, nativeMemoryTag, MemoryTag.MMAP_DEFAULT);
+        this.spillPath = path.concat(tmpFile).$();
+        ff.touch(path);
+    }
+
+    public MemoryCARWSpillable(long pageSize, int maxPages, long diskSpillThreshold, Path path, int nativeMemoryTag, int spillMemoryTag){
+        this(pageSize, maxPages, diskSpillThreshold, nativeMemoryTag, spillMemoryTag);
+        this.spillPath = path.concat(tmpFile).$();
+        ff.touch(path);
+    }
+
+
+
+    MemoryCARWSpillable(long pageSize, int maxPages, long diskSpillThreshold, int nativeMemoryTag, int spillMemoryTag){
+        this.pageSize = Numbers.ceilPow2(pageSize);
+        this.extendSegmentMsb = Numbers.msb(pageSize);
+        this.maxPages = maxPages;
+        this.diskSpillThreshold = Numbers.ceilPow2(diskSpillThreshold);
+        assert diskSpillThreshold > pageSize;
+        this.nativeMemoryTag = nativeMemoryTag;
+        this.spillMemoryTag = spillMemoryTag;
+        this.mem = initCARW();
+        this.isSpilled = false;
+        this.tmpFile = UUID.randomUUID() + ".tmp";
```

Comment:
The file name and path should be provided externally, i.e. from the `FastMap`. This class shouldn't bother with file name generation.